base/loading.jl

```diff
@@ -288,6 +288,7 @@ function reload(name::AbstractString)
         error("use `include` instead of `reload` to load source files")
     else
         # reload("Package") is ok
+        delete!(loaded_modules, Symbol(name))
```

Comment:
This seems like it can be losing gc-roots. Although in practice, it will perhaps often get rooted in the Tuple type cache, I'm not sure if that'll be entirely sufficient to make sure the contained types can get garbage-collected while there is still instances of their objects around.

Reply:
Agreed, we can just keep an old-modules list.